corehq/ex-submodules/pillowtop/pillow/interface.py

```diff
@@ -348,6 +350,16 @@ def __record_change_metric_in_datadog(self, metric, change, processor=None, proc
             if processing_time:
                 datadog_histogram('commcare.change_feed.processing_time', processing_time, tags=tags)
 
+    @staticmethod
+    def _deduplicate_changes(changes_chunk):
+        seen = set()
+        unique = []
+        for change in reversed(changes_chunk):
+            if change.id not in seen:
+                unique.insert(0, change)
```

Comment:
Just re-ran for list of 12000 changes:
```
(u'list append', 5.9798829555511475)
(u'list insert', 5.876524925231934)
(u'dque', 5.824309825897217)
(u'orderd dict', 6.03476095199585)
```
Very little difference between the first 3.

Reply:
Interesting, and kudos for measuring. I wonder how big N would need to be before `list.insert` times began to suffer?

Reply:
Hmm, my timings are quite different from yours. I only measured `insert` vs `append`.

The implementations:
```py
    @staticmethod
    def _deduplicate_changes(changes_chunk):
        seen = set()
        unique = []
        for change in reversed(changes_chunk):
            if change.id not in seen:
                unique.insert(0, change)
                seen.add(change.id)
        return unique

    @staticmethod
    def _deduplicate_changes_append(changes_chunk):
        seen = set()
        unique = []
        for change in reversed(changes_chunk):
            if change.id not in seen:
                unique.append(change)
                seen.add(change.id)
        unique.reverse()
        return unique
```

The test:
```py
    TEST_SIZE = 10000
    NUMBER = 1000
    def test_deduplicate_changes(self):
        changes = [Change(n, 'a') for n in xrange(1, self.TEST_SIZE)] + [
            Change(1, 'b'),
            Change(2, 'b'),
        ]
        expected = [(n, 'a') for n in xrange(3, self.TEST_SIZE)] + [
            (1, 'b'),
            (2, 'b'),
        ]

        # make sure we get the expected output
        deduped = PillowBase._deduplicate_changes_append(changes)
        self.assertEqual(
            [(change.id, change.sequence_id) for change in deduped],
            expected,
        )
        deduped_append = PillowBase._deduplicate_changes_append(changes)
        self.assertEqual(
            [(change.id, change.sequence_id) for change in deduped_append],
            expected,
        )

        import timeit
        print("len(changes)=%s NUMBER=%s" % (len(changes), self.NUMBER))
        print((
            "insert",
            timeit.timeit(lambda : PillowBase._deduplicate_changes(changes), number=self.NUMBER)
        ))
        print((
            "append",
            timeit.timeit(lambda : PillowBase._deduplicate_changes_append(changes), number=self.NUMBER)
        ))
        assert 0, 'fail test to show stdout'
```

Results:
```
len(changes)=101 NUMBER=10000
(u'insert', 0.7732532024383545)
(u'append', 0.5302181243896484)

len(changes)=1001 NUMBER=10000
(u'insert', 11.633922100067139)
(u'append', 5.534351110458374)

len(changes)=10001 NUMBER=1000
(u'insert', 56.22914695739746)
(u'append', 8.600734949111938)
```

In any case, for small chunk sizes (~100) it does not matter that much.

Reply:
Verified your results and also discovered why mine were the same. Most of the items in my test list were excluded because they were duplicates (reminder to validate any assumptions about test data).

Thanks for working through this.

Reply:
updated to use `append`